### ISSUE TITLE
Add support for IMAP IDLE in `imaplib`

### ISSUE THREAD

**Issue by 271e9923-fd98-4048-bbcb-a950400e630c** [python/cpython#55454]
BPO | [11245](https://bugs.python.org/issue11245)
--- | :---
Nosy | @warsaw, @ericvsmith, @bitdancer, @vadmium, @mitya57, @soltysh, @ankostis, @jdek
Dependencies | <li>bpo-18921: In imaplib, cached capabilities may be out of date after login</li>
Files | <li>[imapidle.patch](https://bugs.python.org/file27400/imapidle.patch "Uploaded as text/plain at 2012-10-03.13:59:31 by nafur")</li><li>[imapidle.patch](https://bugs.python.org/file37555/imapidle.patch "Uploaded as text/plain at 2014-12-29.15:26:16 by nafur"): New patch</li><li>[0001-Add-IDLE-support-to-imaplib.patch](https://bugs.python.org/file48790/0001-Add-IDLE-support-to-imaplib.patch "Uploaded as text/plain at 2019-12-19.12:58:09 by @jdek"): IMAP IDLE patch for imaplib</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2011-02-18.20:36:10.487>
labels = ['type-feature', 'library', 'expert-email', '3.9']
title = 'Implementation of IMAP IDLE in imaplib?'
updated_at = <Date 2019-12-19.12:58:09.754>
user = 'https://bugs.python.org/ShayRojansky'
```

bugs.python.org fields:
```python
activity = <Date 2019-12-19.12:58:09.754>
actor = 'jdek'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)', 'email']
creation = <Date 2011-02-18.20:36:10.487>
creator = 'Shay.Rojansky'
dependencies = ['18921']
files = ['27400', '37555', '48790']
hgrepos = []
issue_num = 11245
keywords = ['patch']
message_count = 24.0
messages = ['128814', '129405', '129407', '171885', '171889', '172365', '172383', '202149', '202342', '202345', '233176', '235972', '236167', '245204', '245246', '245249', '245252', '245253', '245255', '245284', '246235', '246236', '293222', '358678']
nosy_count = 17.0
nosy_names = ['barry', 'pierslauder', 'eric.smith', 'piers', 'r.david.murray', 'Shay.Rojansky', 'martin.panter', 'mitya57', 'maciej.szulik', 'nafur', 'dveeden', 'Malina', 'F.Malina', 'ankostis', 'equaeghe', 'ohreally', 'jdek']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'test needed'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue11245'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-122542

**Comment by 384a3c8b-ad2e-4654-ab8d-fa5204db7a0d** [comment]
Before merging `imaplib2` please consider making proper use of the Python's standard `logging` module.

**Comment by trentbuck** [comment]
FYI,

Here is a bare-minimum version that Works For Me (so far) with python 3.9 and dovecot 1.2.9 (don't ask).
This is meant to replace a crappy cron job `* * * * * nobody curl https://⋯/receiveEmail.php >/dev/null`.
I'm think sure this will eventually OOM, as I never explicitly reap the "still here" continuation responses. 

```
#!/usr/bin/python3 

import imaplib
import logging


# Enable IDLE support. 
if 'IDLE' in imaplib.Commands:

    # in the unlikely event this feature is fixed upstream... 
    IMAP4_SSL = imaplib.IMAP4_SSL

else:

    class IMAP4_SSL_plus_IDLE(imaplib.IMAP4_SSL):

        def idle(self):
            if 'IDLE' not in self.capabilities:
                raise self.error('Server does not support IDLE')
            idle_tag = self._command('IDLE')  # start idling 
            self._get_response()
            while line := self._get_line():
                if line.endswith(b'EXISTS'):
                    self.send(b'DONE' + imaplib.CRLF)
                    return self._command_complete('IDLE', idle_tag)

    imaplib.Commands['IDLE'] = ('AUTH', 'SELECTED')
    IMAP4_SSL = IMAP4_SSL_plus_IDLE


with IMAP4_SSL(⋯) as conn:
    conn.login(user=⋯, password=⋯)
    conn.select(mailbox='INBOX', readonly=True)
    conn.debug = 100            # DEBUGGING 
    while True:
        resp = requests.get('https://⋯/receiveEmail.php')
        resp.raise_for_status()
        logging.debug('HTTP GET said %s', resp.text)
        resp = conn.idle()
        logging.debug('IMAP IDLE said %s', resp)
```

**Comment by serhiy-storchaka** [comment]
I am sorry, but I am not sure the interface proposed in #122542 is the best possible interface. There are some issues with it:

* Using the iterator protocol means that the polling is blocking. When we are blocked on waiting a response from the server, there is no way to interrupt waiting. It means that an interactive application which displays an IMAP4 folder status at runtime cannot be responsible to user input. We need a non-blocking operation to poll a response if there is one and return immediately if there are no responses. We need also a blocking operation with timeout and a way to interrupt thread before expiration of the timeout.
* Using the context manager protocol means that DONE is only sent after leaving the `with` block. After client sent DONE, server can still send status updates. What happens to them? You cannot use `Idler` after leaving the `with` block. We need a method to send DONE and read status updates that came after this.
* Server can send status updates not only for IDLE command, but for other commands. It is common to use the NOOP command for this. There is a large similarity between NOOP and IDLE commands, but is is not seen in the proposed interface. You can read status updates from `untagged_responces` after the NOOP command, although this is not thread-safe. It would be nice to have unified way to get status updates for IDLE and other commands. It would be nice if it is thread-safe or there is a clear way to use the IMAP4 client in thread-safe way.

Let's look what other solutions provide. Interesting, there were two different third-party packages named `imaplib2`, both were proposed for inclusion in the stdlib.

[Piers Lauder's `imaplib2`](https://github.com/jazzband/imaplib2/) was discussed above. It is available on PyPI. I has a simple blocking `idle()` method which waits for first server response or expiration of the timeout. Then it send DONE and wait for tagged response for the IDLE command. You can read status updates from `untagged_responces` after this. It may be not very efficient, because you need to send IDLE again, but it fits perfectly in the line with other commands. This package is compatible with `imaplib`. The main disadvantage of this package is that it runs internal threads. But I think that there may be possibility to implement it without running internal thread. Instead the user will need to run external threads to utilize the IMAP4 concurrency features. `imaplib` will still need to use locks and events to make its methods thread-safe.

[Maxim Khitrov's `imaplib2`](https://github.com/mxk/py-imaplib2/) was discussed in [Python-ideas](https://mail.python.org/archives/list/python-ideas@python.org/thread/C4TVEYL5IBESQQPPS5GBR7WFBXCLQMZ2/). It provides incompatible interface, in particularly concurrent `fetch()`. Its interface for IDLE is similar to the one proposed in #122542, with support for the context manager and iterator protocols, but it also has the `poll()` method.

Looking at support in other programming languages, in Perl, [support for IDLE](https://metacpan.org/dist/Mail-IMAPClient/view/lib/Mail/IMAPClient.pod#idle) is provided by three methods -- `idle`, `idle_data` and `done`. `idle` is non-blocking, it simply sends the IDLE command, `idle_data` is non-blocking or blocking with timeout, it polls server response. `done` sends DONE.

**Comment by gvanrossum** [comment]
This landed, so closing.
